Closes #18

amatsonkali opened the pull request:
Saving xpubs on a custom pallet and linking it to the account's identity

amatsonkali committed:
added nbv-storage custom pallet

amatsonkali committed:
Basic xpub storage functionality

Todo: create errors/events and make some tests

amatsonkali committed:
adds nbv basic error handling

amatsonkali committed:
Adds basic unit tests

Code optimizations and transactional behavior

amatsonkali committed:
storage reworked

amatsonkali committed:
Added new error handling

amatsonkali committed:
Added more unit tests

Cleaned up a bit, and add even more validations

amatsonkali committed:
Code dependencies cleanup

mgravitt commented on pallets/nbv-storage/src/lib.rs:
```diff
@@ -0,0 +1,273 @@
+		/// Xpub and hash stored. Linked to an account identity
+		XPubStored([u8 ; 32], T::AccountId),
+		/// Removed Xpub previously linked to the account
+		XpubRemoved(T::AccountId),
```
This is a tiny complaint, but I think the `P` should be capitalized on both events. Or lowercase, but either way it should be consistent, I think

mgravitt commented on pallets/nbv-storage/src/lib.rs:
```diff
@@ -0,0 +1,273 @@
+
+			let mut xpub_tuple = (pallet_identity::Data::None, pallet_identity::Data::None);
+			let updated_fields = identity.info.additional.clone().try_mutate(
+				|addittional_fields|{
```
Two many `t`'s :)

mgravitt replied:
Sorry, this one wasn't clear... I meant there were too many `t`'s in the word `addittional_fields` 

```
let updated_fields = identity.info.additional.clone().try_mutate(
				|addittional_fields|{
                    xpub_id_field = addittional_fields.remove(xpub_index);
					()
				}
```

It should be spelled `additional_fields`.. Sorry for being so picky :)

mgravitt commented on pallets/nbv-storage/src/lib.rs:
```diff
@@ -0,0 +1,273 @@
+		_,
+		Blake2_256,
+		T::AccountId,
+		BoundedVec<u8, T::PSBTMaxLen>,
```
Does Bitcoin Core (or BlueWallet) have a maximum PSBT size? @sebastianmontero

sebastianmontero replied:
I don't think so, it depends on the number of inputs and outputs, I guess we could set a limit?

sebastianmontero replied:
@amatsonkali , @3yekn shouldn't we have a relationship between the psbts and a specific proposal or trx?

amatsonkali replied:
Its a convention within substrate to have an upper limit for all byte-like data that must be stored on-chain.The length limit for each PSBT is currently set at 2048 bytes on the current runtime config.
Got it, I'll add the relationship requirement on the issue #20

mgravitt commented on pallets/nbv-storage/src/lib.rs:
```diff
@@ -0,0 +1,273 @@
+	impl<T: Config> Pallet<T> {
+		/// Use with caution
+		pub fn remove_xpub_from_pallet_storage(who : T::AccountId){
+			// No error can be propagated from the remove functions
```
Why can no error be propagated? Should there be a check that sees if it exists first, and if not, propagate/return an error?

mgravitt commented on pallets/nbv-storage/src/tests.rs:
```diff
@@ -0,0 +1,126 @@
+	}
+}
+
+fn dummy_psbt() -> BoundedVec<u8, PSBTMaxLen> {
```
Is there also a dummy `xpub` value that is saved?

mgravitt replied:
This may be enough of the xpub for the app; perhaps it depends on the `derivation path` assumptions made in BDK. Some xpubs show a fingerprint and the derivation path.  

Here's an example from the `bdk-cli`:

```sh
bdk-cli key derive --xprv tprv8ZgxMBicQKsPdp21UgwHvn6ayoSbctppbD2SwChwNzkXLQKf6fYb6Kjtfww2muuTuZLRVpEV1Y4UbJYdwRnxBb1EPeqxGZmRbSo5NLXYWxc --path "m/84'/1'/0'/0"
```

It produces this output: 
```json
{
  "xprv": "[621c051d/84'/1'/0'/0]tprv8hnMza7pP8tvATh7qb2eUdxwuekBTtueSJFZkji9pJpu2LqdrT6ijHqe87SE2hXkt4Q6aU3JaTKTjLvVX9K4gupjJFN2WFs8sLafYNQySf5/*",
  "xpub": "[621c051d/84'/1'/0'/0]tpubDEUQ8zA4XWab3viujEhEt3d4UgG7dE6Z1brM3FkTEadHrq6QUqvJunTWJH6suxgGsFqspT9AMhASFTj531BVb145NKL2ZzyYGDNQaF2733U/*"
}
```
I don't know if the backend will expect the string beginning `xpub6A...` or it will expect the string beginning `[621c05...`

@sebastianmontero - do you know? 

@amatsonkali the fingerprint is always 8 bytes. I guess the derivation path could vary if the `/0` values go to `/10` or `/100`.  I guess that this can be handled by the XPubLen constant, but the UI <-> pallet <-> bdk services will need to all be aligned on the structure of the `xpub` string

mgravitt commented on pallets/nbv-storage/README.md:
```diff
@@ -0,0 +1 @@
+License: Unlicense
```
In the pallet README, it would be helpful to add example `polkadot-js-api` CLI commands that submit extrinsics and also read the pallet state. It is helpful for JS developers to be able to integrate the functionality with UI screens.

amatsonkali replied:
Got it. I made all the suggested changes, including the documentation

mgravitt reviewed (changes requested):
I added some small questions/comments and also requested some additional info on the README. Looks really good

amatsonkali committed:
PR changes

- P inconsistencies solved
- Too many T's solved
- PSBT issue updated
- Error propagation upon xpub removal
- Unit test refactored
- Readme partially updated

amatsonkali committed:
Updates NBV storage documentation

mgravitt commented:
Documentation in the pallet README is awesome! It has both the CLI walkthrough and JS !

amatsonkali committed:
Minor changes

- fixes some typos (t's)
- Expands xpub length
- Adds complete dummy data: fingerprint+derivation path+xpub on unit tests
